### ISSUE TITLE
Time Series Interpolation is wrong

### ISSUE THREAD

**Issue by oricou** [pandas-dev/pandas#21351]
#### Problem description

```python
dates = pd.date_range('2016-08-28', periods=5, freq='21H') 
ts1 = pd.Series(np.arange(5), dates)
ts1
```

```
2016-08-28 00:00:00    0
2016-08-28 21:00:00    1
2016-08-29 18:00:00    2
2016-08-30 15:00:00    3
2016-08-31 12:00:00    4
Freq: 21H, dtype: int64
```

```python
ts1.resample('15H').interpolate(method='time')
```

```
2016-08-28 00:00:00    0.0
2016-08-28 15:00:00    0.0
2016-08-29 06:00:00    0.0
2016-08-29 21:00:00    0.0
2016-08-30 12:00:00    0.0
2016-08-31 03:00:00    0.0
Freq: 15H, dtype: float64
```

The answer whould not be always 0.

Note that without `method='time'` the result is the same.

#### It can look like ok

If I chose another frequency in the beginning, 20H instead of 21H, then it is ok except the last value:

```python
dates = pd.date_range('2016-08-28', periods=5, freq='20H')
ts1 = pd.Series(np.arange(5), dates)
ts1.resample('15H').interpolate(method='time')
```

```
2016-08-28 00:00:00    0.00
2016-08-28 15:00:00    0.75
2016-08-29 06:00:00    1.50
2016-08-29 21:00:00    2.25
2016-08-30 12:00:00    3.00
2016-08-31 03:00:00    3.00
Freq: 15H, dtype: float64
```
My interpretation is that it puts NaN everywhere except on times it has data that is `2016-08-28 00:00:00` and `2016-08-30 12:00:00` and after it does a linear interpolation. 

My example is bad because I used range(4) which is linear. If I set values to `0 9 9 3 9` then the interpolation gives the same result which is totaly wrong now.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.10.0-38-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.utf8
LOCALE: fr_FR.UTF-8

pandas: 0.23.0
pytest: 3.5.0
pip: 10.0.1
setuptools: 39.0.1
Cython: None
numpy: 1.14.2
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.3.1
sphinx: None
patsy: 0.5.0
dateutil: 2.7.2
pytz: 2018.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.1.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: None
pymysql: None
psycopg2: 2.7.4 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

**Comment by oricou** [comment]
I got it ! It is so stupid, resample does the work before any interpolation and put zero (or Nan) everywhere. Then interpolation cannot find the answer...

**Comment by oricou** [comment]
If interpolation is made after resampling, it makes sens it cannot find the answer, therefore I tried to do both in the same time:

```python
ts1.resample('15H', fill_method='interpolate')
```
and the  answer is:
```
FutureWarning: fill_method is deprecated to .resample()
the new syntax is .resample(...).interpolate()
```

Damned, so there is something wrong. How can we resample and interpolate?

**Comment by oricou** [comment]
And here is the result we should get with original frequency of 21H:

```python
df1 = pd.DataFrame(ts1)
df2 = pd.DataFrame(ts1.asfreq('15H'))
tmp = pd.concat([df1, df2]).sort_index().interpolate(method='time').drop_duplicates()
pd.merge(tmp, df2, how='inner', on=0, left_index=True, right_index=True)
```

```
2016-08-28 00:00:00	0.000000
2016-08-28 15:00:00	0.714286
2016-08-29 06:00:00	1.428571
2016-08-29 21:00:00	2.142857
2016-08-30 12:00:00	2.857143
2016-08-31 03:00:00	3.571429
```

This is, IMHO, what should return
```python
ts1.resample('15H').interpolate(method='time')
```

**Comment by gfyoung** [comment]
@oricou : Indeed, that is odd!  Investigation and PR are welcome!

**Comment by jbrockmendel** [comment]
Possibly related to #19291, #19199, might be something salvageable from the aborted PR.

**Comment by oricou** [comment]
2 years later still the same bug... I suggested a correction which might be not optimal but at least it gives the good result, what make it so long to correct that bug?

**Comment by oricou** [comment]
BTW I don't understand the Missing-Data flag. You just have to run the 4 first cells at the top of the thread.

**Comment by jreback** [comment]
@oricou you are welcome to have a look 

there are 3500 open issues and pandas is all volunteer

the missing data tag is that this is about missing data itself

**Comment by oricou** [comment]
ok, I will try

I still don't understand the Missing Data tag. What is expected ?

**Comment by jreback** [comment]
> 
> 
> 
> I still don't understand the Missing Data tag. What is expected ?

nothing - this is a tag that indicates this issue is about nans

**Comment by JulianWgs** [comment]
This issue describes the same issue: https://github.com/pandas-dev/pandas/issues/21418

The interpolation is only done for values for which are in the index. This produces really unexpected and hard to interpret results. As a quick fix I would propose to add text with this information to the documentation.

**Comment by JulianWgs** [comment]
Here ist the shorted example I could come up with:

```python
import pandas as pd

index = pd.date_range(start="2020-01-01 00:00:0.0", end="2020-01-16 00:00:0.0", freq="3D")
data = (np.linspace(1, 16, len(index)) < 8).astype(float)
ds = pd.Series(index=index, data=data).rename("Original")
pd.concat([ds, ds.resample("5D").interpolate().rename("Resampled")], axis=1)
```


|                     |   Original |   Resampled |
|:--------------------|-----------:|------------:|
| 2020-01-01 00:00:00 |          1 |    1        |
| 2020-01-04 00:00:00 |          1 |  nan        |
| 2020-01-06 00:00:00 |        nan |    0.666667 |
| 2020-01-07 00:00:00 |          1 |  nan        |
| 2020-01-10 00:00:00 |          0 |  nan        |
| 2020-01-11 00:00:00 |        nan |    0.333333 |
| 2020-01-13 00:00:00 |          0 |  nan        |
| 2020-01-16 00:00:00 |          0 |    0        |

The data from the original Series is interpolated with the rows 2020-01-01 and 2020-01-16, because these are the only dates, which are in the original and resampled Series.

**Comment by oricou** [comment]
Why should we do a fix to send a warning when I explain how to do it correctly (see my second message)?

I know I should send a Pull request but I almost done it and I am not conviced it will do any good.

**Comment by JulianWgs** [comment]
I did not want to attack you, I was just trying to help. Sorry. This issue has been open for almost 2 years, how could I know that you are almost there delivering a pull request? Could you please link the pull request even if its only a draft (Github now allows to mark Pull Requests as drafts), so people know about your progress?

In the meanwhile I did my own investigation. In my opinion this line is responsible for the issue:
https://github.com/pandas-dev/pandas/blob/94e29822556259efa2e29bdaa37ffb4f6c470bf3/pandas/core/resample.py#L861

This sets all values not in the new index to nan (everything but 2020-01-01 and 2020-01-16). You can see that by calling `.asfreq()` on the resample object.

Instead the code should combine both indices, do the interpolation and drop the old index.

Is this similar to your approach? How did you approach it?

Best regards
Julian

**Comment by oricou** [comment]
> Here ist the shorted example I could come up with:
> 
> ```python
> import pandas as pd
> 
> index = pd.date_range(start="2020-01-01 00:00:0.0", end="2020-01-16 00:00:0.0", freq="3D")
> data = (np.linspace(1, 16, len(index)) < 8).astype(float)
> ds = pd.Series(index=index, data=data).rename("Original")
> pd.concat([ds, ds.resample("5D").interpolate().rename("Resampled")], axis=1)
> ```
> 
> Original	Resampled
> 2020-01-01 00:00:00	1	1
> 2020-01-04 00:00:00	1	nan
> 2020-01-06 00:00:00	nan	0.666667
> 2020-01-07 00:00:00	1	nan
> 2020-01-10 00:00:00	0	nan
> 2020-01-11 00:00:00	nan	0.333333
> 2020-01-13 00:00:00	0	nan
> 2020-01-16 00:00:00	0	0
> The data from the original Series is interpolated with the rows 2020-01-01 and 2020-01-16, because these are the only dates, which are in the original and resampled Series.

The Resampled is wrong, the result should be 1 1 1 1 0 0 0 0 IMHO

**Comment by oricou** [comment]
All you need to do is to add these lines in pandas/pandas/core/resample.py

line 1181
```
        tmp = concat([obj, result]).sort_index().interpolate(method='time')
        result = tmp[result.index]
        result = result[~result.index.duplicated(keep='first')]
```
and in the header (I don't know exactly where):

```
from pandas.core.reshape.concat import concat
```

to get:

```{python}
    def _upsample(self, method, limit=None, fill_value=None):
        """
        Parameters
        ----------
        method : string {'backfill', 'bfill', 'pad',
            'ffill', 'asfreq'} method for upsampling
        limit : int, default None
            Maximum size gap to fill when reindexing
        fill_value : scalar, default None
            Value to use for missing values

        See Also
        --------
        .fillna: Fill NA/NaN values using the specified method.

        """
        self._set_binner()
        if self.axis:
            raise AssertionError("axis must be 0")
        if self._from_selection:
            raise ValueError(
                "Upsampling from level= or on= selection "
                "is not supported, use .set_index(...) "
                "to explicitly set index to datetime-like"
            )

        ax = self.ax
        obj = self._selected_obj
        binner = self.binner
        res_index = self._adjust_binner_for_upsample(binner)

        # if we have the same frequency as our axis, then we are equal sampling
        if (
            limit is None
            and to_offset(ax.inferred_freq) == self.freq
            and len(obj) == len(res_index)
        ):
            result = obj.copy()
            result.index = res_index
        else:
            result = obj.reindex(
                res_index, method=method, limit=limit, fill_value=fill_value
            )

        tmp = concat([obj, result]).sort_index().interpolate(method='time')
        result = tmp[result.index]
        result = result[~result.index.duplicated(keep='first')]
        return self._wrap_result(result)
```

It should be tested !

**Comment by MarcoGorelli** [comment]
@oricou did you try running `pytest pandas/tests/resample` with your changes? I got 122 failed test cases, here's the diff:
```diff
$ git diff
diff --git a/pandas/core/resample.py b/pandas/core/resample.py
index 2308f9edb4..1ca3c6ec18 100644
--- a/pandas/core/resample.py
+++ b/pandas/core/resample.py
@@ -1175,8 +1175,10 @@ class DatetimeIndexResampler(Resampler):
             result = obj.reindex(
                 res_index, method=method, limit=limit, fill_value=fill_value
             )
-
-        result = self._apply_loffset(result)
+        from pandas.core.reshape.concat import concat
+        tmp = concat([obj, result]).sort_index().interpolate(method='time')
+        result = tmp[result.index]
+        result = result[~result.index.duplicated(keep='first')]
         return self._wrap_result(result)
 
     def _wrap_result(self, result):
```

So, that might not quite be the solution

**Comment by oricou** [comment]
Nope... I do it.

**Comment by JulianWgs** [comment]
Please push your changes into a Pull Request to end the my/your machine debate.

Follow these links for more information:
[How to create a pull request](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request)
[How to contribute to pandas](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html)

**Comment by oricou** [comment]
@MarcoGorelli I started the tests and I added the frequency to the result since my last modification broke it. Now it is better but I still have many errors. The first error is not mine I would say:

tests/resample/test_base.py ligne 55 says the result should be:

```
E   [right]: [0.0, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, 1.0, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, 2.0, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, 3.0, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, 4.0, nan, nan, nan, ...]
```

which is wrong from my point of view, the correct answer is the interpolation I provide I think:

```
E   [left]:  [0.0, 0.041666666666666664, 0.08333333333333333, 0.125, 0.16666666666666666, 0.20833333333333331, 0.25, 0.2916666666666667, 0.3333333333333333, 0.375, 0.41666666666666663, 0.4583333333333333, 0.5, 0.5416666666666666, 0.5833333333333334, 0.625, 0.6666666666666666, 0.7083333333333333, 0.75, ...]
```

So what should I do when the tests expect a wrong answer?

**Comment by oricou** [comment]
@JulianWgs Don't expect people to be as much involved as you can. If the deal is "do all the work" or "do nothing, not even a ticket" you will loose many help.

I don't understand you reference to "my/your machine debate" but it seems agressive.

**Comment by MarcoGorelli** [comment]
> So what should I do when the tests expect a wrong answer?

Open a pull request with your changes, change the test you think is wrong, and see if reviewers agree

**Comment by oricou** [comment]
`pytest pandas/tests/resample` still gives 4 errors. Can someone helps me to solve them? Here is my diff:

```python
% git diff    
diff --git a/pandas/core/resample.py b/pandas/core/resample.py
index 2308f9edb4..736ac2308a 100644
--- a/pandas/core/resample.py
+++ b/pandas/core/resample.py
@@ -86,6 +86,8 @@ from pandas.tseries.offsets import (
     Tick,
 )
 
+from pandas.core.reshape.concat import concat
+
 _shared_docs_kwargs: Dict[str, str] = {}
 
 
@@ -858,7 +860,12 @@ class Resampler(BaseGroupBy, ShallowMixin):
         """
         Interpolate values according to different methods.
         """
+        obj = self._selected_obj
+
         result = self._upsample("asfreq")
+        tmp = concat([obj, result]).sort_index().interpolate(method='time')
+        tmp = tmp[result.index]
+        result[...] = tmp[~tmp.index.duplicated(keep='first')]
         return result.interpolate(
             method=method,
             axis=axis,
```

**Comment by oricou** [comment]
`pytest pandas/tests/resample` still gives 2 errors. Can someone help me to solve them? Here is my diff

```python
% git diff   
diff --git a/pandas/core/resample.py b/pandas/core/resample.py
index 2308f9edb4..f682ff07cd 100644
--- a/pandas/core/resample.py
+++ b/pandas/core/resample.py
@@ -86,6 +86,8 @@ from pandas.tseries.offsets import (
     Tick,
 )
 
+from pandas.core.reshape.concat import concat
+
 _shared_docs_kwargs: Dict[str, str] = {}
 
 
@@ -859,6 +861,11 @@ class Resampler(BaseGroupBy, ShallowMixin):
         Interpolate values according to different methods.
         """
         result = self._upsample("asfreq")
+        if type(result.index) is DatetimeIndex:
+            obj = self._selected_obj
+            tmp = concat([obj, result]).sort_index().interpolate(method='time')
+            tmp = tmp[result.index]
+            result[...] = tmp[~tmp.index.duplicated(keep='first')]
         return result.interpolate(
             method=method,
             axis=axis,
```

**Comment by jreback** [comment]
@oricou your best best is to actually open a pull request

**Comment by oricou** [comment]
With one test which does not pass? Ok, let's try. Thanks.

**Comment by Jython1415** [comment]
I am currently looking at this issue and the related pull request (#39886). Just wanted to make sure I'm not overlapping with anyone, however unlikely it may be.

## Related Issues and PRs

- Interpolating `NaT` (mentioned by @jbrockmendel in [comment](https://github.com/pandas-dev/pandas/issues/21351#issuecomment-396469999) above). It may be related? I'm not so sure.
  - "Potentially related" #19199 closed as duplicate of #11701
  - #11701 closed as duplicate of #11312
  - #11312 closed by #51005
- #21418 (mentioned by @JulianWgs and @oricou) 
  - This uses `resample().interpolate(method="linear")` instead of `method="time"`
  - I have not tried to reproduce this issue yet

## To Reproduce

I have been able to replicate the error described in this issue. See the following example:

```python
import pandas as pd
import numpy as np
resample_rule = "18T"
dti = pd.date_range("2023-06-10", periods=3, freq="30T")
original = pd.Series(np.arange(len(dti), dtype=float), dti)
resampled = original.resample(resample_rule).sum() # any func will do
resampled.loc[:] = np.nan                          # reset values
expected = (
    pd.concat([original, resampled])
    .sort_index()
    .interpolate(method="time")
    .drop(
        original.index.drop(resampled.index, errors="ignore"), errors="ignore"
    )
    .drop_duplicates()
)
result = original.resample(resample_rule).interpolate(method="time")
print(f"> original\n{origin}\n\n> expected\n{expected}\n\n> result\n{result}")
```

<details><summary>Output</summary>
<p>

```text
> original
2023-06-10 00:00:00    0.0
2023-06-10 00:30:00    1.0
2023-06-10 01:00:00    2.0
Freq: 30T, dtype: float64

> expected
2023-06-10 00:00:00    0.0
2023-06-10 00:18:00    0.6
2023-06-10 00:36:00    1.2
2023-06-10 00:54:00    1.8
dtype: float64

> result
2023-06-10 00:00:00    0.0
2023-06-10 00:18:00    0.0
2023-06-10 00:36:00    0.0
2023-06-10 00:54:00    0.0
Freq: 18T, dtype: float64
```

</p>
</details>

The interpolation will fail for `"18T"` but will succeed for `"15T"` and `"20T"`. I think this is because of the overlap between the original and resampled indices, thus giving `interpolate` an anchor to work with. However, I haven't really dug into this.

<details><summary>Successful interpolation for <code>"15T"</code> and <code>"20T"</code></summary>
<p>

`resample_rule = "15T"`

```text
> original
2023-06-10 00:00:00    0.0
2023-06-10 00:30:00    1.0
2023-06-10 01:00:00    2.0
Freq: 30T, dtype: float64

> expected
2023-06-10 00:00:00    0.0
2023-06-10 00:15:00    0.5
2023-06-10 00:30:00    1.0
2023-06-10 00:45:00    1.5
2023-06-10 01:00:00    2.0
dtype: float64

> result
2023-06-10 00:00:00    0.0
2023-06-10 00:15:00    0.5
2023-06-10 00:30:00    1.0
2023-06-10 00:45:00    1.5
2023-06-10 01:00:00    2.0
Freq: 15T, dtype: float64
```

`resample_rule = "20T"`

```text
> original
2023-06-10 00:00:00    0.0
2023-06-10 00:30:00    1.0
2023-06-10 01:00:00    2.0
Freq: 30T, dtype: float64

> expected
2023-06-10 00:00:00    0.000000
2023-06-10 00:20:00    0.666667
2023-06-10 00:40:00    1.333333
2023-06-10 01:00:00    2.000000
dtype: float64

> result
2023-06-10 00:00:00    0.000000
2023-06-10 00:20:00    0.666667
2023-06-10 00:40:00    1.333333
2023-06-10 01:00:00    2.00000
```

</p>
</details>

Note that `assert_series_equal` will fail with how I am generating the `expected` Series because `Series.index.freq` is lost at the `concat` step. `assert_series_equal(check_freq=False)` will pass.

**Comment by oricou** [comment]
die 12/06/23, ad 03h44, Joshua Shew ***@***.***> dixit :
> I am currently looking at this issue and the related pull request (#39886). Just wanted to make sure I'm not overlapping with anyone, however unlikely it may be.

You can work on it. You are most welcome!

Olivier.

**Comment by cbpygit** [comment]
@jreback @jbrockmendel @oricou @Jython1415 
I independently became aware of this aspect. I think there is a relatively important issue here. Before I explain, I filed a PR that fixes how resample->interpolate in the case of upsampling works, here: https://github.com/pandas-dev/pandas/pull/56515 ... But some failing test cases made me realize that the unreasonable behavior is treated as the "expected behavior" so far, so that fixing it would probably require a major version bump if we take semantic versioning seriously. Please read on.

The most important aspect in the code is the comments around the examples here: https://github.com/pandas-dev/pandas/blob/main/pandas/core/resample.py#L1064-L1080 ... The case with 400ms resampling causes only 3 of the original data points to be included when the interpolation method is called. See this plot:

<img width="494" alt="image" src="https://github.com/pandas-dev/pandas/assets/19518682/5014c552-60d6-43f6-9e6b-cfd6eacc7cd3">
 
The blue points are the original example data (`series`). The green points are the 400ms resample-interpolate results with the current code. It is obvious that the green points are not a proper upsampling result for the blue reference points. However, the comment in the doc-string declares this as "desired" (or expected) behavior:

  >  Internal reindexing with ``asfreq()`` prior to interpolation leads to an interpolated timeseries on the basis the reindexed timestamps (anchors). Since not all datapoints from original series become anchors, it can lead to misleading interpolation results [...]

However, objectively nobody would "expect" this result, and it is correctly labeled as "misleading" in the comment. The purpose of such interpolation is precisely to either (a) increase the time resolution of a time series using some assumed underlying function (i.e. linear, polynomial, ...) or to (b) align measurement data done at potentially arbitrary timestamps to a fixed grid. In neither case the current implementation leads to the expected results. It only works when the new desired sampling includes all of the known data points _exactly_. The orange points in the plot show you the interpolation result with my fixed code, achieved by first concatenating the non-covered data points with the new resampled index, then doing the interpolation, then discarding the excess data points. I'd argue this is the result most people would expect.

In the edge case, the current behavior leads to completely unreasonable results, such as:
~~~python
index = pd.date_range('2000-01-01 00:01:00', periods=5, freq='2h')
df = pd.Series(np.arange(5.), index)
df.resample('1h').interpolate(method='linear')
~~~
printing
~~~
2000-01-01 00:00:00   NaN
2000-01-01 01:00:00   NaN
2000-01-01 02:00:00   NaN
2000-01-01 03:00:00   NaN
2000-01-01 04:00:00   NaN
2000-01-01 05:00:00   NaN
2000-01-01 06:00:00   NaN
2000-01-01 07:00:00   NaN
2000-01-01 08:00:00   NaN
Freq: H, dtype: float64
~~~
Here, a 1-second offset makes the upsampling lead to all-NaN results.

If I am given green light that it is okay to correct the interpolation behavior in version 2.x, I can fix the test cases which are based on hard-coded wrong expectations.

**Comment by MarcoGorelli** [comment]
with the caveat that I had time to fully digest your message, there is a version 3.0 coming up, so if it's necessary to correct any current wrong behaviour, then this may be a good chance - want to make a PR?

**Comment by cbpygit** [comment]
@MarcoGorelli Thank you for open ears and quick feedback :) Please note a PR is already there: https://github.com/pandas-dev/pandas/pull/56515 ... I can target 3.0 instead (and adjust the implementation accordingly). But I want to stress that the behavior in 2.x is "wrong", so it would be wise to carefully consider whether it could target 2.x to benefit users of pandas 2. What do you think?

**Comment by MarcoGorelli** [comment]
Nice, thanks

I doubt we get it reviewed that fast I'm afraid (also, looks like there's still a test failure?)

**Comment by cbpygit** [comment]
So much for long messages 😄 ... Let me try shorter (other see longer message above for proper details, please):

  - There is [a PR](https://github.com/pandas-dev/pandas/pull/56515) with a working implementation (not ideal, but working in most cases, not breaking things)
  - There are failing tests, because **these tests hard-coded "wrong" expected behavior**
  - --> Ergo, there are 2 options:
    1.  I fix the failing test cases, but this will also change what 2.x declares as expected behavior --> slightly violates sem-ver, but the behavior it fixes cannot possibly be useful for anybody
    2. We leave 2.x what it is, and rather target 3.0

**Comment by MarcoGorelli** [comment]
it's unlikely there's time for `i.`, so I'm suggesting `ii.` - for both cases the currently-wrong test needs updating though

**Comment by cbpygit** [comment]
Okay, what branch do I need to target? (I wasn't able to figure out where the 3.0 code lives)

**Comment by MarcoGorelli** [comment]
just the `main` branch, there'll probably be a 2.2.x branch later this month after 2.2 r.c. goes out (and if there's a 2.3, it'll likely be cut from the 2.2 branch directly)

**Comment by cbpygit** [comment]
@MarcoGorelli okay, but I am targeting `main` now with the PR, does it mean this will work once I fixed the remaining test cases?

**Comment by MarcoGorelli** [comment]
I'd suggest fixing the test cases within the same PR, we can't merge unless it's green
